packages/apollo-server-core/src/plugin/usageReporting/addTraceRequestStats.ts

```diff
@@ -0,0 +1,418 @@
+import {
+  EnumTypeStat,
+  EnumValueStat,
+  IEnumValueStat,
+  IInputFieldStat,
+  InputFieldStat,
+  InputTypeStat,
+  Trace
+} from 'apollo-reporting-protobuf';
+import { VariableValues } from 'apollo-server-types';
+import {
+  DocumentNode,
+  EnumValueNode,
+  getNamedType,
+  GraphQLInputType,
+  GraphQLSchema,
+  GraphQLType,
+  isEnumType,
+  isInputObjectType,
+  isInputType,
+  isListType,
+  isNonNullType,
+  isScalarType,
+  ObjectFieldNode,
+  ObjectValueNode,
+  OperationDefinitionNode,
+  separateOperations,
+  typeFromAST,
+  TypeInfo,
+  visit,
+  visitWithTypeInfo
+} from "graphql";
+import { isCollection, forEach } from 'iterall';
+
+// We would like to inform users about what schema changes are safe, specifically:
+// - Can an input object field be safely removed?
+// - Can an input object field's default value be safely removed?
+// - Can an input object field's type be safely changed to non-nullable? Will
+//   this require the addition of a default value?
+// - Can an enum value be safely removed?
+//
+// To give this insight, we need to know whether an operation is using a given
+// enum value or input object field (and whether it supplies null at least once
+// for that field, and similarily whether it supplies undefined at least once
+// for that field). This isn't extractable from a given operation signature,
+// since signatures hide literals and don't include variable structure (as this
+// information can be highly dynamic and/or sensitive). So for each request, we
+// summarize just the data we need and add it to the trace.
+export function addTraceRequestStats({
+  trace,
+  schema,
+  document,
+  operation,
+  variables,
+}: {
+  trace: Pick<Trace, 'perEnumTypeStat' | 'perInputTypeStat'>;
+  schema: GraphQLSchema;
+  document: DocumentNode;
+  operation: OperationDefinitionNode;
+  variables?: VariableValues;
+}): void {
+  try {
+    // Search the variable values for input object fields and enum values. This
+    // code is adapted from coerceVariableValues() in graphql-js.
+    //
+    // Note that we need to keep track of which variables evalulate to null and
+    // undefined, for cases where an input value in the operation body contains
+    // an input object field that is set to a variable.
+    const nullVariableNames = new Set<string>();
+    const undefinedVariableNames = new Set<string>();
+    for (const varDefinition of operation.variableDefinitions ?? []) {
+      const varName = varDefinition.variable.name.value;
+
+      // TS unfortunately doesn't handle overloads and union types as nicely as
+      // Flow, see https://github.com/microsoft/TypeScript/issues/14107
+      let varType: GraphQLType | undefined;
+      switch (varDefinition.type.kind) {
```

Comment:
Does this need to be in a loop to deal with "list of non-null", etc? (I think there might be a function in graphql-js to do that loop for you.)

Reply:
I don't believe so, it's handled recursively by `typeFromAST()`. The issue is that the type definition for `typeFromAST()` is overloaded, so whatever type is passed into `typeFromAST()` as the second argument needs to match one of those overloads. `varDefinition.type` is specifically a union type though (`TypeNode = NamedTypeNode | ListTypeNode | NonNullTypeNode`), and there is no overload for that specific union type. So instead we use a `switch` statement to break apart the union type, and call the correct overload of `typeFromAST()` within the appropriate `case`. Flow is smarter than this, and will let you call `typeFromAST()` on a union type even if there's not an overload for that type (provided there's an overload for each member of the union type), but TS won't let you do that. That's the only reason this isn't just a single statement `const varType = typeFromAST(schema, varDefinition.type);`.